receiver/k8sclusterreceiver/watcher.go

```diff
@@ -75,39 +74,119 @@ func newResourceWatcher(
+	return rw, nil
 }
 
-func (rw *resourceWatcher) prepareSharedInformerFactory() {
+func (rw *resourceWatcher) prepareSharedInformerFactory() error {
 	factory := informers.NewSharedInformerFactoryWithOptions(rw.client, 0)
 
-	// Add shared informers for each resource type that has to be watched.
-	rw.setupInformers(&corev1.Pod{}, factory.Core().V1().Pods().Informer())
-	rw.setupInformers(&corev1.Node{}, factory.Core().V1().Nodes().Informer())
-	rw.setupInformers(&corev1.Namespace{}, factory.Core().V1().Namespaces().Informer())
-	rw.setupInformers(&corev1.ReplicationController{},
-		factory.Core().V1().ReplicationControllers().Informer(),
-	)
-	rw.setupInformers(&corev1.ResourceQuota{}, factory.Core().V1().ResourceQuotas().Informer())
-	rw.setupInformers(&corev1.Service{}, factory.Core().V1().Services().Informer())
-	rw.setupInformers(&appsv1.DaemonSet{}, factory.Apps().V1().DaemonSets().Informer())
-	rw.setupInformers(&appsv1.Deployment{}, factory.Apps().V1().Deployments().Informer())
-	rw.setupInformers(&appsv1.ReplicaSet{}, factory.Apps().V1().ReplicaSets().Informer())
-	rw.setupInformers(&appsv1.StatefulSet{}, factory.Apps().V1().StatefulSets().Informer())
-	rw.setupInformers(&batchv1.Job{}, factory.Batch().V1().Jobs().Informer())
-	rw.setupInformers(&batchv1.CronJob{}, factory.Batch().V1().CronJobs().Informer())
-	rw.setupInformers(&autoscalingv2beta2.HorizontalPodAutoscaler{},
-		factory.Autoscaling().V2beta2().HorizontalPodAutoscalers().Informer(),
-	)
+	// Map of supported group version kinds by name of a kind.
+	// If none of the group versions are supported by k8s server for a specific kind,
+	// informer for that kind won't be set and a warning message is thrown.
+	// This map should be kept in sync with what can be provided by the supported k8s server versions.
+	// List of group version kinds must be ordered starting from the most recent versions.
+	// Informer will be configured to use the latest version supported by k8s API.
+	supportedKinds := map[string][]schema.GroupVersionKind{
+		"Pod":                     {gvk.Pod},
+		"Node":                    {gvk.Node},
+		"Namespace":               {gvk.Namespace},
+		"ReplicationController":   {gvk.ReplicationController},
+		"ResourceQuota":           {gvk.ResourceQuota},
+		"Service":                 {gvk.Service},
+		"DaemonSet":               {gvk.DaemonSet},
+		"Deployment":              {gvk.Deployment},
+		"ReplicaSet":              {gvk.ReplicaSet},
+		"StatefulSet":             {gvk.StatefulSet},
+		"Job":                     {gvk.Job},
+		"CronJob":                 {gvk.CronJob, gvk.CronJobBeta},
+		"HorizontalPodAutoscaler": {gvk.HorizontalPodAutoscaler},
+	}
+
+kinds:
+	for kind, gvks := range supportedKinds {
+
+		// Setup informer for the latest supported group version
+		for _, gvk := range gvks {
+			supported, err := rw.isKindSupported(gvk)
+			if err != nil {
+				return err
+			}
+			if supported {
+				rw.setupInformerForKind(gvk, factory)
+				continue kinds
```

Comment:
If i'm reading this correctly, if it finds the CronJob v1 GVK, it won't set up an informer on CronJob v1beta1.  Does this mean that in clusters which support both versions (e.g. 1.21) if a user creates a v1beta1 CronJob they won't get any metrics (since we are only watching for v1 CronJobs)?

Reply:
Yes, this is the behavior. v1beta1 CronJob is deprecated in 1.21, so setting up watchers for it throws warnings. I believe It may be confusing to show warnings for users even if they migrated all the cronjob manifests.

I'm open to other opinions here

Reply:
I would prefer watching both versions if both are available.  Warning logs are OK IMO, but for a user that upgrades from v1.20 to v1.21 (only v1beta1 support to v1 and v1beta1 support), they would lose their metrics until they update their manifests.  Even if they knew about this detail, it wouldn't be possible to keep metrics flowing during the version upgrade

Reply:
Sounds reasonable. I update the code